File: 3398.smallest-substring-with-identical-characters-i.cpp
#
# @lc app=leetcode id=3398 lang=cpp
#
# [3398] Smallest Substring With Identical Characters I
#
# @lc code=start
class Solution {
public:
    int minLength(string s, int numOps) {
        int n = s.length();
        
        // Special case: check if we can make it alternating (k=1)
        int flips0 = 0, flips1 = 0;
        for (int i = 0; i < n; i++) {
            if (s[i] != ((i % 2 == 0) ? '0' : '1')) flips0++;
            if (s[i] != ((i % 2 == 0) ? '1' : '0')) flips1++;
        }
        if (min(flips0, flips1) <= numOps) return 1;
        
        // Find all consecutive runs
        vector<int> runs;
        int count = 1;
        for (int i = 1; i < n; i++) {
            if (s[i] == s[i-1]) {
                count++;
            } else {
                runs.push_back(count);
                count = 1;
            }
        }
        runs.push_back(count);
        
        // Binary search on answer (starting from 2 since we checked 1 above)
        int left = 2, right = *max_element(runs.begin(), runs.end());
        int answer = right;
        
        while (left <= right) {
            int mid = (left + right) / 2;
            if (canAchieve(runs, mid, numOps)) {
                answer = mid;
                right = mid - 1;
            } else {
                left = mid + 1;
            }
        }
        
        return answer;
    }
    
    bool canAchieve(vector<int>& runs, int k, int numOps) {
        int flipsNeeded = 0;
        for (int len : runs) {
            if (len > k) {
                flipsNeeded += len / (k + 1);
            }
        }
        return flipsNeeded <= numOps;
    }
};
# @lc code=end

File: 3398.smallest-substring-with-identical-characters-i.java
```java
// @lc app=leetcode id=3398 lang=java
// [3398] Smallest Substring With Identical Characters I
// @lc code=start
class Solution {
    public int minLength(String s, int numOps) {
        int n = s.length();
        int left = 1, right = n;
        while (left < right) {
            int mid = left + (right - left) / 2;
            if (canAchieve(s, numOps, mid)) {
                right = mid;
            } else {
                left = mid + 1;
            }
        }
        return left;
    }

    private boolean canAchieve(String s, int numOps, int k) {
        return check(s, numOps, k, '0') || check(s, numOps, k, '1');
    }

    private boolean check(String s, int numOps, int k, char target) {
        int n = s.length();
        int l = 0;
        int windowFlips = 0;
        for (int r = 0; r < n; ++r) {
            if (s.charAt(r) != target) {
                windowFlips++;
            }
            if (r - l + 1 > k) {
                if (s.charAt(l) != target) {
                    windowFlips--;
                }
                l++;
            }
            if (r - l + 1 == k) {
                if (windowFlips <= numOps) {
                    return true;
                }
            }
        }
        return false;
    }
}
// @lc code=end
```